xarray/tests/test_dataset.py

```diff
@@ -2104,6 +2104,65 @@ def test_stack_unstack(self):
         actual = ds[['b']].stack(z=['x', 'y']).unstack('z')
         assert actual.identical(ds[['b']])
 
+    def test_to_stacked_array_dtype(self):
+        # make a two dimensional dataset
+        x = DataArray(pd.Index(np.r_[:10], name='x'))
+        y = DataArray(pd.Index(np.r_[:20], name='y'))
+        a = x * y
+        b = x * y * y
+
+        feature_dims = ['y']
+        D = xr.Dataset({'a': a, 'b': b.isel(y=0)})
+        y = D.to_stacked_array('features', feature_dims)
+        assert y.indexes['features'].levels[1].dtype == D.y.dtype
+
+    def test_to_stacked_array_to_unstacked_dataset(self):
+        # make a two dimensional dataset
+        x = DataArray(pd.Index(np.r_[:10], name='x'))
+        y = DataArray(pd.Index(np.r_[:20], name='y'))
+        a = x * y
+        b = x * y * y
+        D = xr.Dataset({'a': a, 'b': b})
+
+        feature_dims = ['y']
+        y = D.to_stacked_array('features', feature_dims)\
+            .transpose("x", "features")
+
+        assert y.dims == ('x', 'features')
+
+        x = y.to_unstacked_dataset("features")
+        self.assertDatasetIdentical(D, x)
+
+        # test on just one sample
+        x0 = y[0].to_unstacked_dataset("features")
+        d0 = D.isel(x=0)
+        self.assertDatasetIdentical(d0, x0)
+
+        # test when data is cast as numpy array
+        y0 = y.data[0]
+        coords = (y.coords['features'],)
+        y0 = xr.DataArray(y0, coords)
+        x0 = y0.to_unstacked_dataset('features')
+
+        # test when variables have different dimensionality
+        D = xr.Dataset({'a': a, 'b': b.isel(y=0)})
+        y = D.to_stacked_array('features', feature_dims)
+        x = y.to_unstacked_dataset('features')
+        self.assertDatasetIdentical(D, x)
+
+        # another test
+        ds = D.isel(x=0)
+        ds_flat = ds.to_stacked_array('features', ['y'])
+        ds_comp = ds_flat.to_unstacked_dataset('features')
+        self.assertDatasetIdentical(ds, ds_comp)
+
+        # test for scalar variables
+        a = xr.DataArray(np.r_[:6], dims=('x'), coords={'x': np.r_[:6]})
+        ds = xr.Dataset({'a': a, 'b': 1.0})
+        ds_flat = ds.to_stacked_array('features', ['x'])
+        ds_comp = ds_flat.to_unstacked_dataset('features')
+        self.assertDatasetIdentical(ds, ds_comp)
```

Comment:
Can you break these into a bunch of smaller tests?

Reply:
Sure. I tried to split these tests into more coherent groups